File: src/data-structures-and-algorithms-with-javascript-book/algorithms/insertionSort.js
function insertionSort(arr) {
    const len = arr.length
    let preIndex, current

    for(let i = 1; i < len; ++i) {
        current = arr[i]
        preIndex = i - 1

        while(preIndex >= 0 && arr[preIndex] > current) {
            arr[preIndex + 1] = arr[preIndex]

            --preIndex
        }

        arr[preIndex + 1] = current
    }

    return arr
}


const array = [61, 85, 19, 88, 68, 8, 70, 29]
const sortedArray = insertionSort(array)

console.log(sortedArray) // [ 8, 19, 29, 61, 68, 70, 85, 88 ]

File: src/03/insertionSort/insertionSort.go
package main

import (
    "fmt"
)

func insertionSort(array []int) []int {
    length := len(array)

    for i := 1; i < length; i++ {
        preIndex := i-1
        current := array[i]

        for preIndex >= 0 && current < array[preIndex] {
            array[preIndex+1] = array[preIndex]
            preIndex--
        }

        array[preIndex+1] = current
    }

    return array
}

func main() {
    arr := []int{12,32,52,1,3,56,74,24,8}
    fmt.Println(insertionSort(arr))
}
